<a href="https://colab.research.google.com/github/isamontec/Analitica-de-Negocios/blob/main/Arbol%20de%20decision%20reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **CASO DE ESTUDIO (RETO 2)**

El objetivo  es desarrollar un modelo predictivo que permita pronosticar si las personas incluidas en la base de datos desarrollarán una enfermedad cardíaca. Este problema se enmarca dentro del análisis predictivo en salud, donde permita priorizar pacientes de alto riesgo y optimizar recursos en prevención y diagnóstico.
Las variables del análisis son:

Age (Edad): representa la edad del paciente en años. Es un factor determinante en enfermedades cardiovasculares, ya que el riesgo aumenta con el envejecimiento.

Systolic (Presión sistólica): mide la presión arterial cuando el corazón se contrae. Valores elevados pueden indicar hipertensión, uno de los principales factores de riesgo cardiovascular.

Diastolic (Presión diastólica): corresponde a la presión arterial cuando el corazón está en reposo entre latidos. También es un indicador clave del estado cardiovascular.

BMI (Índice de Masa Corporal): refleja la relación entre peso y estatura, y permite identificar sobrepeso u obesidad, condiciones asociadas a mayor riesgo cardíaco.

0. Se procede con la carga de las librerias de trabajo

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1.Se procede con la carga de los datos de trabajo

In [ ]:
import pandas as pd
nxl='/content/2. HeartDisease.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XD=XDB.iloc[:,[1,2,3,4]]
yd=XDB.iloc[:,0]
display(XD)

,Age,Systolic,Diastolic,BMI
0,44,112,111,17
1,55,128,90,27
2,47,131,94,26
3,31,151,104,17
4,65,148,117,17
...,...,...,...,...
175,43,119,76,25
176,63,107,113,40
177,47,116,61,27
178,70,110,88,24


2. Se procede con la implementacion del modelo de arbol

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd) # Aquí el modelo busca la relacion entrada?/salida

# Y que fue lo que hizo el modelo?

ydp=mar.predict(XD) # Esto es lo que pronóstica el modelo
#Se construye la matriz de confusion
cm=confusion_matrix(yd,ydp)
display(cm)
VN=cm[0,0]; FP=cm[0,1];FN=cm[1,0]; VP=cm[1,1]

#metricas de desempeño
exactitud=(VP+VN)/(VP+VN+FP+FN) #1.comportamiento general
print('La exactitud es:',exactitud)
sensibilidad=VP/(VP+FN) #2.Como se comporta pronosticando PreAprobados
print('La sensibilidad es:',sensibilidad)
especificidad=VN/(VN+FP) #3. Como se comporta frente a los negativos
print('La especificidad es:',especificidad)
precision=VP/(VP+FP) #4. Como se comporta pronosticando PreAprobados
print('La precision es:',precision)
preneg=VN/(VN+FN) #5. Como se comporta pronosticando PreNegados
print('La prenegatividad es:',preneg)

array([[63, 10],
       [20, 87]])

La exactitud es: 0.8333333333333334
La sensibilidad es: 0.8130841121495327
La especificidad es: 0.863013698630137
La precision es: 0.8969072164948454
La prenegatividad es: 0.7590361445783133


3. Despliegue del arbol de decision

In [ ]:
nxl='/content/2. HeartDisease.xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)
XD=XDB.iloc[:,[1,2,3,4]]
yd=XDB.iloc[:,0]
display(XD)
ydp=mar.predict(XD) # Esto es lo que pronóstica el modelo
print(ydp)

,Age,Systolic,Diastolic,BMI
0,33,120,64,33
1,55,115,81,32
2,51,130,70,20
3,65,101,63,18
4,62,109,110,38
5,60,130,69,35
6,61,124,95,36
7,53,150,81,41


[0 0 0 1 1 1 1 1]


In [ ]:
from sklearn.tree import export_graphviz # exporta los datos a un grafico
from pydotplus import graph_from_dot_data # es un graficador
vs=['Age','Systolic','Diastolic','BMI']
dot_data=export_graphviz(mar,feature_names=vs) #exportar de numeros a graficos en pdf
graph=graph_from_dot_data(dot_data) #hacemos el grafico
graph.write_png('arbolReto2.png')

True

#**Analisis de Resultado**

A partir de la matriz de confusión [[63, 10], [20, 87]], se observa que el modelo clasificó correctamente 63 casos negativos y 87 casos positivos, presentando 10 falsos positivos y 20 falsos negativos. Esto indica que el modelo tiene una ligera mayor dificultad en evitar falsos negativos, aunque mantiene un desempeño sólido. La exactitud alcanza un 83.33%, lo que evidencia un comportamiento general robusto y superior al umbral mínimo esperado para modelos de clasificación binaria. La sensibilidad es de 81.30%, mostrando buena capacidad para identificar pacientes con enfermedad cardíaca, mientras que la especificidad es de 86.30%, lo que indica un excelente reconocimiento de pacientes sanos. La precisión alcanza 89.69%, lo que significa que cuando el modelo predice enfermedad, acierta en casi el 90% de los casos. La prenegatividad de 75.90% confirma que las predicciones negativas también son confiables, aunque ligeramente inferiores frente a las positivas.

De acuerdo con el árbol de decisión se destaca un nodo puro el cual (0/35) posee la siguiente regla de decisión:

Si una persona cumple con Age > 54.5 y Age <= 69.5 y Diastolic > 85.5, tendrá una probabilidad de una presencia de enfermedad del 100%, ya que es un  nodo puro gini = 0.0 y una distribución value = [0,35]. Esto significa que en ese segmento específico el modelo diferencia perfectamente los pacientes con enfermedad cardíaca, mostrando una separación completamente pura.


Se destacan en este árbol, además de los nodos puros, los nodos extremos que representan combinaciones de variables con muy alta probabilidad de pertenecer a una clase específica.

Se destaca una segunda regla que logra un porcentaje de clasificación positiva cercano al 96% y la cual posee la siguiente estructura:

Si Age > 54.5 y Age <= 69.5 y Diastolic <= 85.5, el nodo presenta una distribución value = [4,52], lo que implica que 52 de 56 casos pertenecen a la clase positiva. Esta regla muestra cómo en personas mayores, incluso con presión diastólica controlada, el riesgo sigue siendo considerablemente alto.

La tercera regla que logra un porcentaje elevado de clasificación negativa (aproximadamente 78%) posee la siguiente estructura:

Si Age <= 54.5 y BMI <= 20.5 y Systolic <= 115.5, el nodo presenta una distribución value = [7,2], lo que indica predominancia de la clase negativa (ausencia de enfermedad). Esta combinación refleja un perfil clínico de menor riesgo: edad menor, presión sistólica controlada e índice de masa corporal bajo.
